File: src/main/java/eu/xenit/move2alf/core/action/ListAction.java
package eu.xenit.move2alf.core.action;

import java.io.File;
import java.util.Map;

import eu.xenit.move2alf.common.Parameters;
import eu.xenit.move2alf.common.exceptions.Move2AlfException;
import eu.xenit.move2alf.core.Action;
import eu.xenit.move2alf.core.SourceSink;
import eu.xenit.move2alf.core.dto.ConfiguredAction;
import eu.xenit.move2alf.core.dto.ConfiguredSourceSink;

public class ListAction extends Action {

	@Override
	protected void executeImpl(ConfiguredAction configuredAction,
			Map<String, Object> parameterMap) {
		ConfiguredSourceSink sinkConfig = (ConfiguredSourceSink) configuredAction
				.getConfiguredSourceSinkSet().toArray()[0];
		SourceSink sink = getSourceSinkFactory().getObject(
				sinkConfig.getClassName());
		
		String basePath = configuredAction.getParameter(SourceAction.PARAM_PATH);
		if (!basePath.endsWith("/")) {
			basePath = basePath + "/";
		}
		
		if(!basePath.startsWith("/")) {
			basePath = "/" + basePath;
		}

		String relativePath = (String) parameterMap.get(Parameters.PARAM_RELATIVE_PATH);
		relativePath = relativePath.replace("\\", "/");

		if (relativePath.startsWith("/")) {
			relativePath = relativePath.substring(1);
		}

		// add "cm:" in front of each path component
		String remotePath = basePath + relativePath;
		String[] components = remotePath.split("/");
		remotePath = "";
		for (String component : components) {
			if ("".equals(component)) {
				remotePath += "/";
			} else if (component.startsWith("cm:")) {
				remotePath += component + "/";
			} else {
				remotePath += "cm:" + component + "/";
			}
		}
		remotePath = remotePath.substring(0, remotePath.length() - 1);
		
		String name = ((File) parameterMap.get(Parameters.PARAM_FILE)).getName();
		try {
			if (sink.exists(sinkConfig, remotePath, name)) {
				parameterMap.put(Parameters.PARAM_STATUS, Parameters.VALUE_OK);
			} else {
				parameterMap.put(Parameters.PARAM_STATUS, Parameters.VALUE_FAILED);
				parameterMap.put(Parameters.PARAM_ERROR_MESSAGE, "Document not found");
			}
		} catch (Move2AlfException e) {
			parameterMap.put(Parameters.PARAM_STATUS, Parameters.VALUE_FAILED);
			parameterMap.put(Parameters.PARAM_ERROR_MESSAGE, e.getMessage());
		}
	}

	@Override
	public String getCategory() {
		return null;
	}

	@Override
	public String getDescription() {
		return null;
	}

	@Override
	public String getName() {
		return null;
	}

}

File: move2alf-war/src/main/scala/eu/xenit/move2alf/core/action/ListAction.scala
package eu.xenit.move2alf.core.action

import eu.xenit.move2alf.core.simpleaction.data.FileInfo
import eu.xenit.move2alf.common.Parameters
import java.io.File
import eu.xenit.move2alf.core.action.messages.ListMessage

/**
 * User: Thijs Lemmens (tlemmens@xenit.eu)
 * Date: 7/2/13
 * Time: 9:52 AM
 */
class ListAction extends ActionWithDestination[FileInfo, Boolean]{
  private var path: String = null
  def setPath(path: String) {
    this.path = path
  }
  protected def executeImpl(fileInfo: FileInfo) {
    val newParameterMap: FileInfo = new FileInfo
    newParameterMap.putAll(fileInfo)
    var basePath: String = path
    if (!basePath.endsWith("/")) {
      basePath = basePath + "/"
    }
    if (!basePath.startsWith("/")) {
      basePath = "/" + basePath
    }
    var relativePath: String = newParameterMap.get(Parameters.PARAM_RELATIVE_PATH).asInstanceOf[String]
    relativePath = relativePath.replace("\\", "/")
    if (relativePath.startsWith("/")) {
      relativePath = relativePath.substring(1)
    }
    // add "cm:" in front of each path component
    var remotePath: String = basePath + relativePath
    val components: Array[String] = remotePath.split("/")
    remotePath = ""
    for (component <- components) {
      if ("" == component) {
        remotePath += "/"
      }
      else if (component.startsWith("cm:")) {
        remotePath += component + "/"
      }
      else {
        remotePath += "cm:" + component + "/"
      }
    }
    remotePath = remotePath.substring(0, remotePath.length - 1)
    val name: String = newParameterMap.get(Parameters.PARAM_NAME).asInstanceOf[String]

    sendTaskToDestination(fileInfo, new ListMessage(remotePath, name), result => {
      if(result) {
        fileInfo.put(Parameters.PARAM_STATUS, Parameters.VALUE_OK)
        sendMessage(fileInfo)
      } else handleError(fileInfo, "This file is not in the repository.")
    })
  }
}

object ListAction{
  val PARAM_PATH: String = "path"
}